File: web-sites-hub/frontend-docs/apple-ds-core-docs/10-union-find/union_find.cpp
/**
 * 并查集实现
 */

#include <iostream>
#include <vector>
#include <numeric>

using namespace std;

class UnionFind {
private:
    vector<int> parent;
    vector<int> rank;
    int count;  // 连通分量数

public:
    UnionFind(int n) : parent(n), rank(n, 0), count(n) {
        iota(parent.begin(), parent.end(), 0);  // 初始化为0,1,2,...
    }
    
    // 查找根节点（路径压缩）
    int find(int x) {
        if (parent[x] != x) {
            parent[x] = find(parent[x]);
        }
        return parent[x];
    }
    
    // 合并两个集合（按秩合并）
    bool unite(int x, int y) {
        int px = find(x);
        int py = find(y);
        
        if (px == py) {
            return false;
        }
        
        if (rank[px] < rank[py]) {
            parent[px] = py;
        } else if (rank[px] > rank[py]) {
            parent[py] = px;
        } else {
            parent[py] = px;
            rank[px]++;
        }
        
        count--;
        return true;
    }
    
    // 判断是否连通
    bool connected(int x, int y) {
        return find(x) == find(y);
    }
    
    // 获取连通分量数
    int getCount() const {
        return count;
    }
};

// 岛屿数量
int numIslands(vector<vector<char>>& grid) {
    if (grid.empty()) return 0;
    
    int m = grid.size();
    int n = grid[0].size();
    UnionFind uf(m * n);
    int zeros = 0;
    
    for (int i = 0; i < m; i++) {
        for (int j = 0; j < n; j++) {
            if (grid[i][j] == '0') {
                zeros++;
            } else {
                // 向右和向下连接
                if (j + 1 < n && grid[i][j+1] == '1') {
                    uf.unite(i * n + j, i * n + j + 1);
                }
                if (i + 1 < m && grid[i+1][j] == '1') {
                    uf.unite(i * n + j, (i + 1) * n + j);
                }
            }
        }
    }
    
    return uf.getCount() - zeros;
}

// 朋友圈数量
int findCircleNum(const vector<vector<int>>& isConnected) {
    int n = isConnected.size();
    UnionFind uf(n);
    
    for (int i = 0; i < n; i++) {
        for (int j = i + 1; j < n; j++) {
            if (isConnected[i][j] == 1) {
                uf.unite(i, j);
            }
        }
    }
    
    return uf.getCount();
}

// 冗余连接
vector<int> findRedundantConnection(const vector<vector<int>>& edges) {
    int n = edges.size();
    UnionFind uf(n + 1);
    
    for (const auto& edge : edges) {
        if (!uf.unite(edge[0], edge[1])) {
            return edge;
        }
    }
    
    return {};
}

int main() {
    cout << "=== 并查集演示 ===" << endl << endl;
    
    UnionFind uf(10);
    
    cout << "初始连通分量数: " << uf.getCount() << endl;
    
    uf.unite(1, 2);
    uf.unite(2, 3);
    uf.unite(4, 5);
    
    cout << "合并后连通分量数: " << uf.getCount() << endl;
    cout << "1和3连通: " << (uf.connected(1, 3) ? "是" : "否") << endl;
    cout << "1和4连通: " << (uf.connected(1, 4) ? "是" : "否") << endl;
    
    uf.unite(3, 4);
    cout << "\n合并3和4后" << endl;
    cout << "连通分量数: " << uf.getCount() << endl;
    cout << "1和5连通: " << (uf.connected(1, 5) ? "是" : "否") << endl << endl;
    
    // 朋友圈
    cout << "朋友圈问题:" << endl;
    vector<vector<int>> isConnected = {
        {1, 1, 0},
        {1, 1, 0},
        {0, 0, 1}
    };
    cout << "  朋友圈数量: " << findCircleNum(isConnected) << endl;
    
    return 0;
}

File: web-sites-hub/frontend-docs/apple-ds-core-docs/10-union-find/union_find.py
"""
并查集实现
"""

class UnionFind:
    """并查集（路径压缩 + 按秩合并）"""
    
    def __init__(self, n):
        self.parent = list(range(n))
        self.rank = [0] * n
        self.count = n  # 连通分量数
    
    def find(self, x):
        """查找根节点（路径压缩）"""
        if self.parent[x] != x:
            self.parent[x] = self.find(self.parent[x])
        return self.parent[x]
    
    def union(self, x, y):
        """合并两个集合（按秩合并）"""
        px, py = self.find(x), self.find(y)
        
        if px == py:
            return False
        
        if self.rank[px] < self.rank[py]:
            self.parent[px] = py
        elif self.rank[px] > self.rank[py]:
            self.parent[py] = px
        else:
            self.parent[py] = px
            self.rank[px] += 1
        
        self.count -= 1
        return True
    
    def connected(self, x, y):
        """判断是否连通"""
        return self.find(x) == self.find(y)
    
    def get_count(self):
        """获取连通分量数"""
        return self.count


# ========== 应用示例 ==========

def num_islands(grid):
    """岛屿数量"""
    if not grid:
        return 0
    
    m, n = len(grid), len(grid[0])
    uf = UnionFind(m * n)
    zeros = 0
    
    for i in range(m):
        for j in range(n):
            if grid[i][j] == '0':
                zeros += 1
            else:
                for di, dj in [(0, 1), (1, 0)]:
                    ni, nj = i + di, j + dj
                    if 0 <= ni < m and 0 <= nj < n and grid[ni][nj] == '1':
                        uf.union(i * n + j, ni * n + nj)
    
    return uf.get_count() - zeros


def find_circle_num(is_connected):
    """朋友圈数量"""
    n = len(is_connected)
    uf = UnionFind(n)
    
    for i in range(n):
        for j in range(i + 1, n):
            if is_connected[i][j] == 1:
                uf.union(i, j)
    
    return uf.get_count()


def find_redundant_connection(edges):
    """找冗余连接"""
    n = len(edges)
    uf = UnionFind(n + 1)
    
    for u, v in edges:
        if not uf.union(u, v):
            return [u, v]
    
    return []


def demo():
    """演示并查集"""
    print("=== 并查集演示 ===\n")
    
    uf = UnionFind(10)
    
    print(f"初始连通分量数: {uf.get_count()}")
    
    # 合并
    uf.union(1, 2)
    uf.union(2, 3)
    uf.union(4, 5)
    
    print(f"合并后连通分量数: {uf.get_count()}")
    print(f"1和3连通: {uf.connected(1, 3)}")
    print(f"1和4连通: {uf.connected(1, 4)}")
    
    uf.union(3, 4)
    print(f"\n合并3和4后")
    print(f"连通分量数: {uf.get_count()}")
    print(f"1和5连通: {uf.connected(1, 5)}")
    
    # 朋友圈问题
    print("\n=== 朋友圈问题 ===\n")
    is_connected = [
        [1, 1, 0],
        [1, 1, 0],
        [0, 0, 1]
    ]
    print(f"朋友圈矩阵: {is_connected}")
    print(f"朋友圈数量: {find_circle_num(is_connected)}")


if __name__ == '__main__':
    demo()

